Framework/API/src/DetectorInfo.cpp

```diff
@@ -67,48 +81,65 @@ bool DetectorInfo::isMasked(const size_t index) const {
  */
 double DetectorInfo::l2(const size_t index) const {
   if (!isMonitor(index))
-    return getDetector(index).getDistance(getSample());
+    return position(index).distance(samplePosition());
   else
-    return getDetector(index).getDistance(getSource()) - l1();
+    return position(index).distance(sourcePosition()) - l1();
 }
 
 /// Returns 2 theta (scattering angle w.r.t. to beam direction).
 double DetectorInfo::twoTheta(const size_t index) const {
-  const Kernel::V3D samplePos = samplePosition();
-  const Kernel::V3D beamLine = samplePos - sourcePosition();
+  if (isMonitor(index))
+    throw std::logic_error(
+        "Two theta (scattering angle) is not defined for monitors.");
+
+  const auto samplePos = samplePosition();
+  const auto beamLine = samplePos - sourcePosition();
 
   if (beamLine.nullVector()) {
```

Comment:
Yes, that would be good. I think allowing the source and sample to be at the same position should generally be allowed, and the direction vector should be obtained from `Beamline`. I am not sure if anything in Mantid might rely on a beam that is not along an axis though.